README.md
# kuma-zipkin

[![996.icu](https://img.shields.io/badge/link-996.icu-red.svg)](https://996.icu)

基于`zipkin-js`二次封装的适配`koa`的中间件，更改了原有日志打印格式和逻辑。

## 使用文档

待补充

[更新日志](CHANGELOG.md)

## 许可证（License）

[Anti-996 License](LICENSE)


CHANGELOG.md
# 更新日志

---

## 0.2.0 - 2022-02-22

### 新增

- 新增`validationError`错误类，当抛出错误类型为该类时，其错误提示优先级最高，目前只在koa中间件中有效

## 0.1.6 - 2022-02-18

### 修复

- `response`缺少换行符bug

## 0.1.5 - 2022-02-18

### 修改

- `response`返回限制为500个字符，有需要时再做配置项调整

## 0.1.4 - 2021-09-18

### 修复

- `axios`在`tracer`未传入时引发的问题

## 0.1.3 - 2021-09-17

### 修改

- 修改日志输出格式
- 修改id输出规则

## 0.1.2 - 2021-08-31

### 新增

- 新增`setHead`配置项

## 0.1.1 - 2021-08-17

### 修复

- 修复文档错误

## 0.1.0 - 2021-08-17

### 新增

- 完成`koa`中间件功能
- 完成`axios`组件功能


src/axios.js
const axios = require("axios");
const { InetAddress, ExplicitContext, ConsoleRecorder, Tracer } = require("zipkin");
const { now: zipkinNow } = require("zipkin/src/time");

const isJson = (it) => {
  try {
    JSON.stringify(it);
  } catch (error) {
    return false;
  }
  return true;
};

const queryText = (data, type) => {
  const isJ = isJson(data);
  if (data === undefined || (isJ && !Object.keys(data).length)) return "";
  if (isJ) return `\n${type}: ${JSON.stringify(data)}`;
  return `\n${type}: ${data.toString()}`;
};

const key2LowerCase = (json = {}) => {
  const out = {};
  for (const key in json) {
    if (Object.hasOwnProperty.call(json, key)) {
      out[key.toLowerCase()] = json[key];
    }
  }
  return out;
};

module.exports = (_cfg = {}) => {
  const {
    remoteServiceName,
    serviceName,
    ctx: { tracer, logger, ip },
    requestLog,
    responseLog,
    ...other
  } = _cfg;
  const { traceId, spanId, parentSpanId } =
    tracer?.id ??
    (() => {
      const ctxImpl = new ExplicitContext();
      const recorder = new ConsoleRecorder();
      return (new Tracer({
        recorder,
        ctxImpl,
        localServiceName: "",
      })).id;
    })();
  const Axios = axios.create({
    ...other,
  });
  const host = (InetAddress.getLocalAddress() + "").match(
    /\d{0,3}\.\d{0,3}\.\d{0,3}\.\d{0,3}/
  )[0];
  // 请求拦截
  Axios.interceptors.request.use((config) => {
    requestLog && requestLog instanceof Function
      ? requestLog({
          traceId,
          spanId,
          parentSpanId,
          method: config.method.toUpperCase(),
          url: config.url,
          ip,
          host,
          query: config.params,
          data: config.data,
          remoteServiceName,
          config,
        })
      : logger.info(
          `[${serviceName ?? ''},${traceId},${spanId},${parentSpanId.toString()}]\n${config.method.toUpperCase()} ${
            config.url
          }\norigin: [from ${ip} to ${host}]${queryText(
            config.params,
            "query"
          )}${queryText(
            config.data,
            "data"
          )}\nsource: ${remoteServiceName} 请求\n`
        );
    if (traceId && spanId) {
      config.headers["X-B3-TraceId"] = traceId;
      config.headers["X-B3-SpanId"] = spanId;
      config.headers["X-B3-ParentSpanId"] = parentSpanId.toString();
    }
    config.startTime = zipkinNow();
    return config;
  });
  // 响应拦截
  Axios.interceptors.response.use((config) => {
    const {
      "x-b3-traceid": _traceId,
      "x-b3-spanid": _spanId,
      "x-b3-parentspanid": _parentSpanId,
      "content-type": type,
    } = key2LowerCase(config.headers ?? {});
    const data = queryText(config?.data, "data");
    const resTraceId = _traceId ?? traceId;
    const resSpanId = _spanId ?? spanId;
    const resParentSpanId = _parentSpanId ?? parentSpanId.toString();
    responseLog && responseLog instanceof Function
      ? responseLog({
          status: config.status,
          traceId: resTraceId,
          spanId: resSpanId,
          parentSpanId: resParentSpanId,
          method: config.request.method,
          url: config.request.path,
          time: zipkinNow() - config.config.startTime,
          type,
          data,
          remoteServiceName,
          config,
        })
      : logger[
          [200, 204].some((item) => item === config.status) ? "info" : "error"
        ](
          `[${serviceName ?? ''},${resTraceId},${resSpanId},${resParentSpanId}]\n${
            config.request.method
          } ${config.request.path}\ntime: ${
            (zipkinNow() - config.config.startTime) / 1000
          }ms\nstatus: ${config.status} <${type ?? "undefined"}>${
            data.length > 500
              ? `${data.substr(0, 500)} <data长度为${
                  data.length
                }，只保留500字符>`
              : data
          }\ndst: ${remoteServiceName} 响应\n`
        );
    return config;
  });
  return Axios;
};


src/koaMiddleware.js
const {
  option: { Some, None },
  Instrumentation,
  InetAddress,
  Tracer,
  ConsoleRecorder,
  ExplicitContext,
} = require("zipkin");
const validationError = require('./validationError.js');

const parseRequestUrl = require("zipkin/src/parseUrl");
const { now, hrtime } = require("zipkin/src/time");
const isJson = (it) => {
  try {
    JSON.stringify(it);
  } catch (error) {
    return false;
  }
  return true;
};
const isFunction = (it) => {
  return Object.prototype.toString.call(it) == "[object Function]";
};
const ctxImpl = new ExplicitContext();
const recorder = new ConsoleRecorder();
const tracer = new Tracer({
  recorder,
  ctxImpl,
  localServiceName: "",
});

class KumaInstrumentation extends Instrumentation.HttpServer {
  constructor(props) {
    super(props);
  }
  recordRequest(method, requestUrl, readHeader) {
    this._createIdFromHeaders(readHeader).ifPresent((id) =>
      this.tracer.setId(id)
    );
    const { id } = this.tracer;
    const { path } = parseRequestUrl(requestUrl);
    const host = ((this.host || InetAddress.getLocalAddress()) + "").match(
      /\d{0,3}\.\d{0,3}\.\d{0,3}\.\d{0,3}/
    )[0];

    return { id, path, method: method.toUpperCase(), host };
  }
}

/**
 * @brief 基于zipkin-js二次封装的koa插件
 * @param {string} serviceName 服务名
 * @param {number} port 端口
 * @param {Object} console 日志输出工具
 * @param {Array} ignore 需要忽略的日志
 * @param {Boolean} header 是否在浏览器头写入id
 * @return {ZipkinKoaMiddleware}
 */
module.exports = function koaMiddleware({
  serviceName = "",
  port = 0,
  console = global.console,
  ignore = [],
  requestLog,
  responseLog,
  setHead = true,
}) {
  const instrumentation = new KumaInstrumentation({
    tracer,
    serviceName,
    port,
  });

  /**
   * @method
   * @typedef {function} ZipkinKoaMiddleware
   * @param {Object} ctx
   * @param {function()} next
   */
  return function zipkinKoaMiddleware(ctx, next) {
    function readHeader(header) {
      const val = ctx.request.headers[header.toLowerCase()];
      if (val != null) {
        return new Some(val);
      } else {
        return None;
      }
    }
    return tracer.scoped(() => {
      const _method = ctx.request.method.toUpperCase();
      const { id, path, method, host } = instrumentation.recordRequest(
        _method,
        ctx.request.href,
        readHeader
      );
      const _ignore = (Array.isArray(ignore) ? ignore : [ignore]).some((item) =>
        isFunction(item)
          ? item({ path, method, ip: ctx.ip, host })
          : ((path.match(item) ?? [])[0] ?? "").length
      );
      id.startTime = now();
      id.startTimeMicro = now(id.startTime, hrtime());

      Object.defineProperty(ctx, "tracer", {
        configurable: false,
        get: () => tracer,
      });

      Object.defineProperty(ctx.request, "_trace_id", {
        configurable: false,
        get: () => id,
      });
      Object.defineProperty(ctx.request, "_trace_path", {
        configurable: false,
        get: () => path,
      });
      Object.defineProperty(ctx.request, "_trace_method", {
        configurable: false,
        get: () => method,
      });
      Object.defineProperty(ctx.request, "_trace_host", {
        configurable: false,
        get: () => host,
      });
      const queryText = (data, type) => {
        const isJ = isJson(data);
        if (data === undefined || (isJ && !Object.keys(data).length)) return "";
        if (isJ) return `\n${type}: ${JSON.stringify(data)}`;
        return `\n${type}: ${data.toString()}`;
      };
      const { traceId, spanId, parentSpanId } = id;

      !_ignore &&
        console.info(
          requestLog && requestLog instanceof Function
            ? requestLog({
              traceId,
              spanId,
              parentSpanId,
              method,
              url: path,
              ip: ctx.ip,
              host,
              query: ctx.query,
              data: ctx.request.body,
              params: ctx.params,
              serviceName,
              ctx,
              id,
            })
            : `[${serviceName},${traceId},${spanId},${parentSpanId}]\n${(
              method + ""
            ).toUpperCase()} ${path}\norigin: [from ${ctx.ip
            } to ${host}]${queryText(ctx.query, "query")}${queryText(
              ctx.request.body,
              "data"
            )}${queryText(ctx.params, "params")}\nsource: 客户端 请求\n`
        );

      const recordResponse = () => {
        const resBody = ["application/json", "text/plain"].some(
          (item) => item === ctx.type
        )
          ? `data: ${JSON.stringify(ctx.body)}\n`
          : "";
        tracer.letId(id, () => {
          // support koa-route and koa-router
          const matchedPath = ctx.routePath || ctx._matchedRoute || ctx.path;
          !_ignore &&
            console[
              [200, 204].some((item) => item === ctx.status) ? "info" : "error"
            ](
              responseLog && responseLog instanceof Function
                ? responseLog({
                  status: ctx.status,
                  traceId,
                  spanId,
                  parentSpanId,
                  method,
                  url: matchedPath,
                  time: now() - id.startTime,
                  type: ctx.type,
                  data: ctx.body,
                  serviceName,
                  ctx,
                  id,
                })
                : `[${serviceName},${traceId},${spanId},${parentSpanId}]\n${(
                  method + ""
                ).toUpperCase()} ${matchedPath}\ntime: ${(now() - id.startTime) / 1000
                }ms\nstatus: ${ctx.status} <${ctx.type
                }>\n${resBody.length > 500
                  ? `${resBody.substr(0, 500)} <data长度为${resBody.length
                  }，只保留500字符>`
                  : resBody
                }\ndst: 客户端 响应\n`
            );
          if (setHead) {
            ctx.set("X-B3-TraceId", traceId);
            ctx.set("X-B3-SpanId", spanId);
            ctx.set("X-B3-ParentSpanId", parentSpanId);
          }
        });
      };

      ctx.console = {};
      ["error", "info", "warn", "debug"].map((item) => {
        ctx.console[item] = (...args) => {
          ctx.logger[item](
            `[traceId=${traceId ?? ""}, spanId=${spanId ?? ""}, parentSpanId=${parentSpanId.toString() ?? ""
            }]\n`,
            ...args
          );
        };
      });
      // 错误处理
      ctx.errorFn = (error, message, status = 500, print = true) => {
        print && ctx.console.error(error);
        ctx.status = error.isAxiosError ? error.response.status : status;
        ctx.body = {
          message:
            error instanceof validationError ? error.message :
              (message ||
                (error.isAxiosError
                  ? error.response?.data?.message
                  : error.message || "系统错误")),
        };
      };

      return next().then(recordResponse).catch(recordResponse);
    });
  };
};
